lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1146 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! # lightning-net
+//!
+//! A socket handling library for those using rust-lightning within a
+//! synchronous, multi-threaded runtime.
+//!
+//! Whereas `lightning-net-tokio` manages reading and writing to peers using
+//! Futures and Tokio tasks, this library uses dedicated blocking threads. While
+//! this does result in a small amount of performance overhead, the
+//! absence of any Tokio `net` features means that this library can successfully
+//! compile to the `x86_64-fortanix-unknown-sgx` (Fortanix EDP) target.
+//! Compiling for SGX was the primary motivation for writing this crate, but
+//! this crate can also be useful for those who simply want to use
+//! rust-lightning without Tokio or async Rust.
+//!
+//! The primary entrypoints into this crate are `initiate_outbound()` and
+//! `handle_connection()`. See their individual docs for details.
+//!
+//! ## `std` limitations of EDP
+//!
+//! Compiling for Fortanix EDP comes with additional limitations, however. This
+//! crate purposefully avoids the use of:
+//!
+//! - `std::time::Instant::now`
+//! - `std::time::Instant::elapsed`
+//! - `std::time::SystemTime::now`
+//! - `std::time::SystemTime::elapsed`
+//! - `std::thread::sleep`
+//! - `std::thread::sleep_ms`
+//! - `std::thread::park_timeout`
+//! - `std::thread::park_timeout_ms`
+//! - `std::net::TcpStream::shutdown`
+//! - `std::net::TcpStream::connect_timeout`
+//! - `std::net::TcpStream::set_read_timeout`
+//! - `std::net::TcpStream::set_write_timeout`
+//! - `std::net::TcpStream::set_nodelay`
+//! - `std::net::TcpStream::set_ttl`
+//! - `std::net::TcpStream::set_nonblocking`
+//! - `std::net::TcpListener::set_nodelay`
+//! - `std::net::TcpListener::set_only_v6`
+//! - `std::net::TcpListener::set_nonblocking`
+//!
+//! These functions have varying degrees of compatibility with Fortanix EDP.
+//! See the [EDP docs](https://edp.fortanix.com/docs/concepts/rust-std/) for
+//! more information on what Rust features can and cannot be used within SGX.
```

Comment:
These are relevant though? Even after implementing a `SgxTcpStream`, the rest of the shared code needs to avoid using these constructs, otherwise the code will panic within SGX. Mentioning the EDP target as a goal makes it clear to future programmers why the module has these additional constraints

Reply:
how are we planning to "swap in" the `SgxTcpStream`? a `trait Socket` that mimics a small subset of the `TcpStream` interface?

Reply:
Something like that. It needs to implement `Read`, `Write`, `Clone`, and `shutdown()` at least. I figure it would be best to add in the generics in a later PR once we've implemented the custom usercalls and have a good idea how our SGX socket works. The eventual `SgxTcpStream` doesn't have to be upstreamed, but IMO the comments about restricted functions should stay in.

Reply:
Perhaps I could just remove every line that starts with `std::net` since the concrete `TcpStream` will be swapped out with our custom `Socket` anyway

Reply:
i think the other functions are also unlikely to be used if we have a `Socket` trait, since the trait would make it clear what the expected interface looks like.